### PR TITLE
OCIO base support

### PR DESCRIPTION
**Pull request by bernardkwok**

### CODE REVIEW

File: source/MaterialXGenShader/OCIOColorManagementSystem.cpp [hunk 1]
```diff
@@ -0,0 +1,634 @@
...
+{
+    return std::make_shared<OCIOSourceCodeNode>();
+}
+
+void OCIOSourceCodeNode::initialize(const InterfaceElement& element, GenContext& context)
+{
+    SourceCodeNode::initialize(element, context);
+
+    // Cache a uniform block
+    for (auto nodeInput : element.getInputs())
```
Comment (kwokcb) [inline source/MaterialXGenShader/OCIOColorManagementSystem.cpp:556]: Init just caches information to use when the node is called for later calls to emit code, create uniforms etc.

File: source/MaterialXGenShader/OCIOColorManagementSystem.cpp [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+    if (!_cmUniforms.empty())
+    {
+        ShaderStage& stage = shader.getStage(Stage::PIXEL);
+        stage.createUniformBlock(ColorManagementSystem::COLOR_MANAGEMENT_UNIFORMS);
+        VariableBlock& uniformBlock = stage.getUniformBlock(ColorManagementSystem::COLOR_MANAGEMENT_UNIFORMS);
+
+        for (size_t i = 0; i < _cmUniforms.size(); ++i)
+        {
+            ShaderPort* u = const_cast<ShaderPort*>(_cmUniforms[i]);
+            uniformBlock.add(u->getSelf());
```
Comment (kwokcb) [inline source/MaterialXGenShader/OCIOColorManagementSystem.cpp:587]: This is where we add uniforms to the appropriate uniform block. Global shader gen logic will determine how to format and emit these. e.g. for some languages we need to create specific resources blocks of uniforms based on type.

File: source/MaterialXGenShader/OCIOColorManagementSystem.h [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+  private:
+    OCIOInformation*  _ocioInfo;
+    FilePath          _configFile;
+    string            _target;
+};
+
+
+/// Extending the SourceCodeNode with requirements for OCIO.
+class MX_GENSHADER_API OCIOSourceCodeNode : public SourceCodeNode
```
Comment (kwokcb) [inline source/MaterialXGenShader/OCIOColorManagementSystem.h:85]: The OCIO transform node is a special derivation of the base source code where the code and signature are dynamically generated.

File: source/MaterialXGenShader/ShaderGraph.cpp [hunk 4]
```diff
@@ -276,23 +276,7 @@ void ShaderGraph::addColorTransformNode(ShaderInput* input, const ColorSpaceTran
...
-        shaderInput->setUnit(EMPTY_STRING);
-
-        if (input->isBindInput())
-        {
-            ShaderOutput* oldConnection = input->getConnection();
-            shaderInput->makeConnection(oldConnection);
-        }
-
-        input->makeConnection(colorTransformNodeOutput);
+        colorManagementSystem->connectNodeToShaderInput(this, colorTransformNodePtr.get(), input, context);
```
Comment (kwokcb) [inline source/MaterialXGenShader/ShaderGraph.cpp:279]: Cleanup.

File: source/MaterialXGenShader/ShaderGraph.cpp [hunk 5]
```diff
@@ -310,20 +294,7 @@ void ShaderGraph::addColorTransformNode(ShaderOutput* output, const ColorSpaceTr
...
-        for (ShaderInput* input : inputs)
-        {
-            input->breakConnection();
-            input->makeConnection(colorTransformNodeOutput);
-        }
-
-        // Connect the node to the upstream output
-        ShaderInput* colorTransformNodeInput = colorTransformNode->getInput(0);
-        colorTransformNodeInput->makeConnection(output);
+        colorManagementSystem->connectNodeToShaderOutput(this, colorTransformNodePtr.get(), output, context);
```
Comment (kwokcb) [inline source/MaterialXGenShader/ShaderGraph.cpp:297]: Cleanup.

File: source/MaterialXRender/ImageHandler.cpp [hunk 6]
```diff
@@ -147,6 +147,28 @@ ImagePtr ImageHandler::acquireImage(const FilePath& filePath)
     return _invalidImage;
 }
 
+ImagePtr ImageHandler::createImage(const string& identifier, unsigned int width, unsigned int height,
```
Comment (kwokcb) [inline source/MaterialXRender/ImageHandler.cpp:150]: Add ability to create image from an in memory block.

File: source/MaterialXRenderGlsl/GLTextureHandler.cpp [hunk 7]
```diff
@@ -56,21 +56,23 @@ bool GLTextureHandler::bindImage(ImagePtr image, const ImageSamplingProperties&
     }      
     _boundTextureLocations[textureUnit] = image->getResourceId();
 
+    GLenum targetType = image->getHeight() > 1 ? GL_TEXTURE_2D : GL_TEXTURE_1D;
+
     glActiveTexture(GL_TEXTURE0 + textureUnit);
-    glBindTexture(GL_TEXTURE_2D, image->getResourceId());
+    glBindTexture(targetType, image->getResourceId());
```
Comment (kwokcb) [inline source/MaterialXRenderGlsl/GLTextureHandler.cpp:62]: Generalize code to be able to specify more than 1D textures.

File: source/MaterialXRenderGlsl/GlslProgram.cpp [hunk 8]
```diff
@@ -548,6 +549,104 @@ void GlslProgram::bindTextures(ImageHandlerPtr imageHandler)
...
 
+void GlslProgram::bindColorManagement(ColorManagementSystemPtr cms, ImageHandlerPtr imageHandler)
+{
+    if (!cms || !imageHandler)
+    {
+        return;
+    }
+
+    // Handle non-texture uniforms
+    ColorManagementResourceMapPtr uniformItems = cms->getResource(ColorManagementSystem::ResourceType::UNIFORM);
```
Comment (kwokcb) [inline source/MaterialXRenderGlsl/GlslProgram.cpp:560]: Resources cached when OCIO transforms are created are bound here.

File: source/MaterialXRenderGlsl/GlslProgram.cpp [hunk 9]
```diff
@@ -548,6 +549,104 @@ void GlslProgram::bindTextures(ImageHandlerPtr imageHandler)
...
+
+                // Create an new image if needed and bind it to the appropriate location
+                // Note that if the image with the uniform name already exists then it will be
+                // returned from the image cache and reused.
+                if (hasUniform(uniformName))
+                {
+                    ImagePtr uniformImage = imageHandler->getImage(uniformName);
+                    if (!uniformImage)
+                    {
+                        uniformImage = imageHandler->createImage(uniformName,
```
Comment (kwokcb) [inline source/MaterialXRenderGlsl/GlslProgram.cpp:604]: Textures (LUT) are dynamically created.

File: source/MaterialXTest/README.md [hunk 10]
```diff
@@ -58,6 +58,10 @@ Per language tests will scan MaterialX files in the test suite for input Element
 If rendering tests are enabled via the build options then code for each Element tested will be compiled and rendered if the appropriate backend support is available.
 - `GLSL`:
     - Will execute on a Windows machine which supports OpenGL 4.0 or above.
+    - The OCIO library can be used during code generation instead of the default system provided the following is set up: 
+        - The MATERIALX_BUILD_OCIO build option is set to be ON. 
+        - The MATERIALX_OCIO_DIR build option is set to point to the root of the OCIO installation.
+        - For unit tests: A valid OCIO configuration file is specified for the `colorManagementConfigFile` option in the `_options.mtlx` options file.
```
Comment (kwokcb) [inline source/MaterialXTest/README.md:64]: Can set a config file to use in options file.

File: source/MaterialXView/Main.cpp [hunk 11]
```diff
@@ -33,7 +33,10 @@ const std::string options =
 "    --remap [TOKEN1:TOKEN2]        Specify the remapping from one token to another when MaterialX document is loaded\n"
 "    --skip [NAME]                  Specify to skip elements matching the given name attribute\n"
 "    --terminator [STRING]          Specify to enforce the given terminator string for file prefixes\n"
-"    --help                         Display the complete list of command-line options\n";
+#ifdef MATERIALX_BUILD_OCIO
+"    --ocioConfigFile [STRING]      Specify a OCIO configuration file to use for color management.\n"
```
Comment (kwokcb) [inline source/MaterialXView/Main.cpp:37]: Can set a config file to use for the viewer.

**Comment by bernardkwok** [comment]
Will proceed with just some updates and no OCIO at this point.

**Pull request closed without merging**